src/factor/factor.rs

```diff
@@ -12,66 +16,174 @@
 
 extern crate getopts;
 extern crate libc;
+extern crate rand;
 
-use std::vec::Vec;
-use std::old_io::BufferedReader;
-use std::old_io::stdio::stdin_raw;
+use numeric::*;
+use prime_table::{P_INVS_U64, NEXT_PRIME};
+use std::cmp::{max,min};
+use std::io::{stdin, BufRead, BufReader, Write};
+use std::num::Wrapping;
+use std::mem::swap;
+use rand::weak_rng;
+use rand::distributions::{Range, IndependentSample};
 
 #[path="../common/util.rs"]
 #[macro_use]
 mod util;
+mod numeric;
+mod prime_table;
 
 static VERSION: &'static str = "1.0.0";
 static NAME: &'static str = "factor";
 
-fn factor(mut num: u64) -> Vec<u64> {
-    let mut ret = Vec::new();
+fn trial_division_slow(mut num: u64, factors: &mut Vec<u64>) {
+    // assumption: this number has already been run through
+    // trial_division, which checks primes from the table.
+    // The first candidate we need to check is NEXT_PRIME.
+    let mut i = NEXT_PRIME;
+    while i * i <= num {
+        while num % i == 0 {
+            num /= i;
+            factors.push(i);
+            if is_prime(num) {
+                factors.push(num);
+                return;
+            }
+            if num < 1 << 63 {
+                // once we're small enough, switch to Pollard's rho
+                return rho_pollard_factor(num, factors);
+            }
+        }
+        i += 2;
+    }
+    if num > 1 {
+        factors.push(num);
+    }
+}
+
+fn rho_pollard_pseudorandom_function(x: u64, a: u64, b: u64, num: u64) -> u64 {
+    (multiply(a, multiply(x, x, num), num) + b) % num
+}
+
+fn gcd(mut a: u64, mut b: u64) -> u64 {
+    while b > 0 {
+        a %= b;
+        swap(&mut a, &mut b);
+    }
+    a
+}
+
+fn rho_pollard_find_divisor(num: u64) -> u64 {
+    let range = Range::new(1, num);
+    let mut rng = rand::weak_rng();
+    let mut x = range.ind_sample(&mut rng);
+    let mut y = x;
+    let mut a = range.ind_sample(&mut rng);
+    let mut b = range.ind_sample(&mut rng);
+
+    loop {
+        x = rho_pollard_pseudorandom_function(x, a, b, num);
+        y = rho_pollard_pseudorandom_function(y, a, b, num);
+        y = rho_pollard_pseudorandom_function(y, a, b, num);
+        let d = gcd(num, max(x,y) - min(x,y));
+        if d == num {
+            // Failure, retry with diffrent function
+            x = range.ind_sample(&mut rng);
+            y = x;
+            a = range.ind_sample(&mut rng);
+            b = range.ind_sample(&mut rng);
+        } else if d > 1 {
+            return d;
+        }
+    }
+}
 
+fn rho_pollard_factor(num: u64, factors: &mut Vec<u64>) {
+    if is_prime(num) {
+        factors.push(num);
+        return;
+    }
+    let divisor = rho_pollard_find_divisor(num);
+    rho_pollard_factor(divisor, factors);
+    rho_pollard_factor(num/divisor, factors);
```

Comment:
There are a few more spots like this that I haven't commented on.  Can you fix those as well?